deltacat/compute/compactor/steps/repartition.py

```diff
@@ -0,0 +1,162 @@
+# which is a major component of the transformation operation thereby reducing the time of completion.
+# Deltacat with Ray can support different partitioning strategies to reduce the data movement either across network or between compute and storage
+# Note that the term partition here is different from the term used in catalog
+# Type of Partition:
+# Range Partition: It assigns rows to partitions based on column values falling within a given range, e.g., repartition(column="last_updated", ranges=['2023-01-01', '2023-02-01', '2023-03-01']), data will be split into 4 files
+# Hash Partition: Hash Partitioning attempts to spread the data evenly across various partitions based on the key, e.g., repartition(column="last_updated", num_partitions=10), data will be split into 10 files evenly
+
+
+class RepartitionType(str, Enum):
+    RANGE = "range"
+    HASH = "hash"
+
+
+def _timed_repartition(
+    annotated_delta: DeltaAnnotated,
+    repartition_type: RepartitionType,
+    repartition_args: dict,
+    destination_partition: Partition,
+    enable_profiler: bool,
+    read_kwargs_provider: Optional[ReadKwargsProvider],
+    repartitioned_file_content_type: ContentType = ContentType.PARQUET,
+    deltacat_storage=unimplemented_deltacat_storage,
+) -> RepartitionResult:
+
+    if repartition_type == RepartitionType.RANGE:
+        column: str = repartition_args["column"]
+        partition_ranges: List = repartition_args["ranges"]
+        task_id = get_current_ray_task_id()
+        worker_id = get_current_ray_worker_id()
+        with memray.Tracker(
+            f"repartition_{worker_id}_{task_id}.bin"
+        ) if enable_profiler else nullcontext():
+            tables: List[pa.Table] = deltacat_storage.download_delta(
+                annotated_delta,
+                storage_type=StorageType.LOCAL,
+                file_reader_kwargs_provider=read_kwargs_provider,
+            )
+            # check if the column exists in the table
+            if not all(column in table.column_names for table in tables):
+                raise ValueError(f"Column {column} does not exist in the table")
+            # given a range [x, y, z], we need to split the table into 4 files, i.e., (-inf, x], (x, y], (y, z], (z, inf)
+            partitioned_tables_list = [[] for _ in range(len(partition_ranges) + 1)]
+            total_record_count = 0
+            col_name_int64 = f"{column}_int64"
+            for table in tables:
+                total_record_count += len(table)
+                table_new = table.add_column(
+                    0,
+                    pa.field(col_name_int64, pa.int64()),
+                    pc.cast(table[column], pa.int64()),
+                )
+                # handle the partition for values less than or equal to the smallest value
+                partitioned_tables_list[0].append(
+                    table_new.filter(
+                        pc.field(col_name_int64) <= pc.scalar(partition_ranges[0])
```

Comment:
This approach iterates over the array M times where M is number of buckets. Isn't it faster to sort the array (in N logN) and iterate through it once to split the array using pyarrow [slice](https://arrow.apache.org/docs/python/generated/pyarrow.Table.html#pyarrow.Table.slice)?